mediachain/transactor/blockchain_follower.py

```diff
@@ -62,82 +64,134 @@ def start(self):
-                            self.block_ref_queue.put('__abort__')
-                            self.caught_up = True
+                            self.catchup_complete.set()
                         else:
                             self.block_ref_queue.put(block_ref)
+                            self.catchup_begin.set()
 
                     self.incoming_event_queue.put(event)
             except AbortionError as e:
-                if self._cancelled and e.code == StatusCode.CANCELLED:
+                if self.cancel_flag.is_set() and e.code == StatusCode.CANCELLED:
                     return
                 else:
                     raise
         with_retry(event_receive_worker, max_retry_attempts=self.max_retry)
 
     def _event_stream(self):
-        # block until catchup thread completes
-        self.catchup_thread.join()
-
-        for block_ref in self.replay_stack:
-            block = self.cache.get(block_ref)
-            entries = block.get('entries', [])
-            for e in entries:
-                e = self.event_map_fn(block_event_to_rpc_event(e))
-                if e is not None:
-                    yield e
-            block_event = Transactor_pb2.JournalEvent()
-            block_event.journalBlockEvent.reference = ref_base58(block_ref)
-            block_event = self.event_map_fn(block_event)
-            if block_event is not None:
-                yield block_event
-
         while True:
+            if self.cancel_flag.is_set():
+                return
+            # wait until catchup process signals that it's complete
+            # this will be immediate if catchup is not in progress
+            self.catchup_complete.wait()
+
+            # get all values from the catchup queue and yield their entries
+            while not self.block_replay_queue.empty():
+                if self.cancel_flag.is_set():
+                    return
+                block_ref = self.block_replay_queue.get()
+                print('Replaying block: {}'.format(block_ref))
+                block = self.cache.get(block_ref)
+                entries = block.get('entries', [])
+                for e in entries:
+                    e = self.event_map_fn(block_event_to_rpc_event(e))
+                    if e is not None:
+                        yield e
+                self.last_known_block_ref = block_ref
+                block_event = Transactor_pb2.JournalEvent()
+                block_event.journalBlockEvent.reference = ref_base58(block_ref)
+                block_event = self.event_map_fn(block_event)
+                if block_event is not None:
+                    yield block_event
+
+            if self.cancel_flag.is_set():
+                return
+            # Try to pull an event off of the incoming event queue
+            # If there's no event received within a second,
+            # sleep for a bit and loop back.
             try:
                 e = self.incoming_event_queue.get(block=False, timeout=1)
```

Comment:
yeah, I just made those numbers up. happy to change them to something that makes sense :) Originally I was just slipping a small sleep interval in there to allow exceptions to be caught, since the blocking Queue.get prevents exceptions from bubbling up.  But now we're also looping back to the top of this loop, which will let us check if a new catchup has been started and process it.

Anyway, yeah, we should use better timeout / sleep numbers :)